ouroboros-consensus-test/bench-storage/Bench/Ouroboros/Consensus/Storage/LedgerDB/HD.hs

```diff
@@ -0,0 +1,412 @@
+  , model    :: !(Model k v)
+  }
+  deriving stock Generic
+  deriving anyclass NFData
+
+{-------------------------------------------------------------------------------
+  Commands
+-------------------------------------------------------------------------------}
+
+type Block (ds :: DiffSeqKind) k v = (SlotNo ds, Diff ds k v)
+
+data Cmd (ds :: DiffSeqKind) k v =
+    Push     (SlotNo ds)     (Diff ds k v)
+  | Flush    Int
+  | Rollback Int             [Block ds k v]
+  | Forward  (Values ds k v) (Keys ds k v)
+  deriving stock Generic
+
+deriving anyclass instance ( NFData (ds k v)
+                           , NFData (Diff ds k v)
+                           , NFData (Values ds k v)
+                           , NFData (Keys ds k v)
+                           , NFData (SlotNo ds)
+                           ) => NFData (Cmd ds k v)
+
+deriving stock instance ( Show (ds k v)
+                        , Show (Diff ds k v)
+                        , Show (Values ds k v)
+                        , Show (Keys ds k v)
+                        , Show (SlotNo ds)
+                        ) => Show (Cmd ds k v)
+
+deriving stock instance ( Eq (ds k v)
+                        , Eq (Diff ds k v)
+                        , Eq (Values ds k v)
+                        , Eq (Keys ds k v)
+                        , Eq (SlotNo ds)
+                        ) => Eq (Cmd ds k v)
+
+{-------------------------------------------------------------------------------
+  Interpreter
+-------------------------------------------------------------------------------}
+
+-- | Interpretation of @Cmd@s as a fold over an initial diff sequence.
+interpret ::
+     forall ds k v.
+     ( IsDiffSeq ds k v
+     , NFData (Diff ds k v)
+     , NFData (Values ds k v)
+     )
+  => ds k v
+  -> [Cmd ds k v]
+  -> ds k v
+interpret = foldl' go
+  where
+    go :: ds k v -> Cmd ds k v -> ds k v
+    go ds = \case
+      Push sl d     -> push ds sl d
+      -- We @'deepseq'@ the total diff of the left part of the flush result,
+      -- because that is the part that would be written to disk, for which it
+      -- must be fully evaluated. In particular, the performance of computing
+      -- the total diff can be different for each diff sequence implementation.
+      Flush n       -> let (l, r) = flush n ds
+                       in  totalDiff l `deepseq` r
+      -- We do not @'deepseq'@ the right part of the rollback result, because we
+      -- are throwing it away in the case of a rollback.
+      Rollback n bs -> let (l, _r) = rollback n ds
+                       in  pushMany l bs
+      -- We @'deepseq'@ the result of forwarding values through a total diff,
+      -- since forwarding can be implementation-specific in (at least) two ways:
+      -- 1. How the total diff is computed.
+      -- 2. How the change to a specific key-value pair is computed.
+      Forward vs ks -> let vs' = forward vs ks ds
+                       in  vs' `deepseq` ds
```

Comment:
Maybe we should comment on the fact that in the actual system, the result of the forward operation will be used (for instance in block application), and we force its evaluation here to account for the costs of the forward operation.